tests/backwards_compatibility_tests/compatibility_test_runner.py

```diff
@@ -0,0 +1,564 @@
+
+        # Step 3: Start to_version container, transferring state
+        start_marqo_to_version_container(to_version, from_version, from_version_volume, to_version_tag)
+        print(f"Started marqo container in to_version {to_version} by transferring state")
+        # Step 4: Run tests
+        run_tests("test", from_version, to_version, "http://localhost:8882")
+        print("Ran tests in test mode")
+    except Exception as e:
+        print(f"Error: {e}, {e.__class__.__name__}, {e.__traceback__}, {e.__traceback__.__class__}, {e.__traceback__.tb_lineno}")
+        raise e
+    finally:
+        # Stop the to_version container (but don't remove it yet)
+        print("Calling stop_marqo_container with" + str(to_version))
+        stop_marqo_container(to_version)
+        # Clean up all containers at the end
+        cleanup_containers()
+        cleanup_volumes()
+
+
+
+def rollback_test(to_version: str, from_version: str, to_version_tag, from_image: Optional[str] = None,
+                  to_image: Optional[str] = None):
+    """
+    Perform a rollback test between two versions of Marqo.
+
+    This function first performs a backwards compatibility test from the from_version to the to_version.
+    Then, it stops the to_version container, starts the from_version container again, and runs tests in test mode.
+
+    Args:
+        to_version (str): The target version of the Marqo container.
+        from_version (str): The source version of the Marqo container.
+        to_version_tag: The specific image tag to use for the to_version container.
+        from_image (Optional[str]): The specific image to use for the from_version container. Defaults to None.
+        to_image (Optional[str]): The specific image to use for the to_version container. Defaults to None.
+    """
+    try:
+        backwards_compatibility_test(from_version, to_version, None, from_image, to_image)
+
+        stop_marqo_container(to_version)
+
+        start_marqo_from_version_container(from_version, None, from_image)
+
+        run_tests("test", from_version, to_version, "http://localhost:8882")
+    finally:
+        # Stop the final container (but don't remove it yet)
+        stop_marqo_container(from_version)
+        # Clean up all containers at the end
+        cleanup_containers()
+
+def run_tests(mode: str, from_version: str, to_version: str, marqo_api: str):
+    print(f"Inside run_tests with arguments mode: {mode}, from_version: {from_version}, to_version: {to_version}")
+
+    if mode == "prepare":
+        # Get all subclasses of `BaseCompatibilityTestCase` that match the `from_version` criterion
+        tests = [test_class for test_class in BaseCompatibilityTestCase.__subclasses__()
+                 if getattr(test_class, 'marqo_from_version', '0') <= from_version]
+
+        for test_class in tests:
+            test_class.setUpClass()
+            test_instance = test_class()
+            test_instance.prepare()
```

Comment:
Do we need to run setUp()?

Reply:
Yes we need to run that, since during prepare mode, we need to create a client so that we can call Marqo to create indexes and add documents. In one of the test cases in the prepare mode we also do search, for which the client will be required. For all of this running setUp() is required

Reply:
Do we have to run setUp() explicitly here? Maybe no because we only run setUp() before each real test?

Reply:
the reasoning behind running it explicitly here is that, even before prepare method is executed, we need to create a client so that we can call Marqo to create indexes and add documents, all of that can be taken care of by setUp(). After setUp() is run - we will create indexes and add documents, and then we will stop the container on which we had created indexes and added documents, and delete that container. 

Now we transfer the underlying state by using volumes, and run a new container, post which we need to run setUp() again so that it creates that client which we can use to query marqo / add more docs / search etc. 

Let me know if this makes sense, we can sync up offline on this.